REGRESSION (283135@main): Headings using `background-clip: text` become rectangles when exported as PDF

smfr opened the pull request:
#### 4d774bdad1260cf75c4b21611a52d14f2d5a3219
<pre>
REGRESSION (283135@main): Headings using `background-clip: text` become rectangles when exported as PDF
<a href="https://bugs.webkit.org/show_bug.cgi?id=285760">https://bugs.webkit.org/show_bug.cgi?id=285760</a>
<a href="https://rdar.apple.com/142698412">rdar://142698412</a>

Reviewed by Tim Horton.

There were a couple of issues with `background-clip: text` that affect printing. First, when printing
skips backgrounds (because of `print-color-adjust: economy` and the user unchecking the Print Backgrounds
checkbox in the print dialog), `background-clip: text` did not show at all because it&apos;s painted as a background.
Fix this by changing the `layerClip` from `FillBox::Text` to `FillBox::BorderBox` in `BackgroundPainter::paintFillLayer()`,
and making the text black in `computeTextPaintStyle()`; the spec[1] allows us to make these adjustments for `economy`.
This is tested by a layout test.

Second, when printing or generating a PDF when including backgrounds, i.e. for `print-color-adjust: exact` or with
the Print Backgrounds checkbox checked, the SourceIn blend mode is not supported, which resulted in a solid background
and missing text. Fix this by falling back to clipping to an image buffer for PDF contexts, in `BackgroundPainter::paintFillLayer()`.
This is tested by a new API test.

[1] <a href="https://drafts.csswg.org/css-color-adjust/#propdef-print-color-adjust">https://drafts.csswg.org/css-color-adjust/#propdef-print-color-adjust</a>

* LayoutTests/platform/gtk/TestExpectations:
* LayoutTests/platform/wpe/TestExpectations:
* LayoutTests/printing/background-clip-text-expected.html: Added.
* LayoutTests/printing/background-clip-text.html: Added.
* Source/WebCore/rendering/BackgroundPainter.cpp:
(WebCore::BackgroundPainter::paintFillLayer const):
* Source/WebCore/rendering/StyledMarkedText.cpp:
(WebCore::StyledMarkedText::computeStyleForUnmarkedMarkedText):
* Source/WebCore/rendering/TextPaintStyle.cpp:
(WebCore::computeTextPaintStyle):
* Source/WebCore/rendering/TextPaintStyle.h:
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::hasEntirelyFixedBackground const):
(WebCore::RenderStyle::hasAnyBackgroundClipText const):
(WebCore::allLayersAreFixed): Deleted.
* Source/WebCore/rendering/style/RenderStyle.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/DrawingToPDF.mm:
(TestWebKitAPI::TEST(DrawingToPDF, BackgroundClipText)):

Canonical link: <a href="https://commits.webkit.org/288753@main">https://commits.webkit.org/288753@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/332f1e6ed00226b8b81046330d4563f58149c8de

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/84291 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/3913 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/38596 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/89369 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/35300 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/86376 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/4000 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/11889 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/65557 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/23393 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/87337 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/3001 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/76589 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/45850 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/2952 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/30819 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/34349 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/73852 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/31587 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/90749 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/11558 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/8395 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/74008 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/11785 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/72414 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/73210 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18120 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/17532 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/15976 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/2923 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/11510 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/16986 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/11359 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/14835 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/13132 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebCore/rendering/BackgroundPainter.cpp:
```diff
@@ -327,18 +331,32 @@ void BackgroundPainter::paintFillLayer(const Color& color, const FillLayer& bgLa
 
-        context.setCompositeOperation(CompositeOperator::SourceIn);
+        if (context.renderingMode() == RenderingMode::PDFDocument) {
+            // CG doesn't support some compositing operations when printing, so clip using an imageBuffer instead.
```
Sad to see mention of CG here, and also workarounds for CG here.

hortont424 replied:
Can it at least be a "GraphicsContext::needs/supportsXYZ" and move the mention of CG in there?

smfr committed:
REGRESSION (283135@main): Headings using `background-clip: text` become rectangles when exported as PDF
https://bugs.webkit.org/show_bug.cgi?id=285760
rdar://142698412

Reviewed by Tim Horton.

There were a couple of issues with `background-clip: text` that affect printing. First, when printing
skips backgrounds (because of `print-color-adjust: economy` and the user unchecking the Print Backgrounds
checkbox in the print dialog), `background-clip: text` did not show at all because it's painted as a background.
Fix this by changing the `layerClip` from `FillBox::Text` to `FillBox::BorderBox` in `BackgroundPainter::paintFillLayer()`,
and making the text black in `computeTextPaintStyle()`; the spec[1] allows us to make these adjustments for `economy`.
This is tested by a layout test.

Second, when printing or generating a PDF when including backgrounds, i.e. for `print-color-adjust: exact` or with
the Print Backgrounds checkbox checked, the SourceIn blend mode is not supported, which resulted in a solid background
and missing text. Fix this by falling back to clipping to an image buffer for PDF contexts, in `BackgroundPainter::paintFillLayer()`.
This is tested by a new API test.

[1] https://drafts.csswg.org/css-color-adjust/#propdef-print-color-adjust

* LayoutTests/platform/gtk/TestExpectations:
* LayoutTests/platform/wpe/TestExpectations:
* LayoutTests/printing/background-clip-text-expected.html: Added.
* LayoutTests/printing/background-clip-text.html: Added.
* Source/WebCore/rendering/BackgroundPainter.cpp:
(WebCore::BackgroundPainter::paintFillLayer const):
* Source/WebCore/rendering/StyledMarkedText.cpp:
(WebCore::StyledMarkedText::computeStyleForUnmarkedMarkedText):
* Source/WebCore/rendering/TextPaintStyle.cpp:
(WebCore::computeTextPaintStyle):
* Source/WebCore/rendering/TextPaintStyle.h:
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::hasEntirelyFixedBackground const):
(WebCore::RenderStyle::hasAnyBackgroundClipText const):
(WebCore::allLayersAreFixed): Deleted.
* Source/WebCore/rendering/style/RenderStyle.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/DrawingToPDF.mm:
(TestWebKitAPI::TEST(DrawingToPDF, BackgroundClipText)):

Canonical link: https://commits.webkit.org/288753@main
